### PR TITLE
Refactor Manifest Classification

### PR DESCRIPTION
**Pull request by 0xOmarA**

### CODE REVIEW

**Review by dhedey — COMMENTED** [review]
Just posting a temporary review of these comments I had sitting around.  I haven't looked through manifest_analysis or access_rule yet

File: crates/sbor-json/src/common/address.rs [hunk 1]
```diff
@@ -34,7 +34,7 @@ impl Display for SerializableNodeId {
             crate::utils::network_definition_from_network_id(self.1);
         let bech32_encoder = AddressBech32Encoder::new(&network_definition);
         bech32_encoder
-            .encode_to_fmt(f, &self.0.0)
+            .encode_to_fmt(f, &self.0 .0)
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/sbor-json/src/common/address.rs:37]: Not really, I think the stable formatter prefers `0 .0` than `0.0` :)

File: rustfmt.toml [hunk 2]
```diff
@@ -1,5 +1 @@
-version = "Two"
-max_width = 80
-comment_width = 80
-unstable_features = true
-imports_granularity = "Module"
\ No newline at end of file
+max_width=80
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply rustfmt.toml:1]: I disagree with this, why would we want to align the style guidelines and formatting between the toolkit and Scrypto?

[the remark this answers is not in the document]
Comment (dhedey) [reply rustfmt.toml:1]: It's standard practice to align style guidelines across a team.

[the remark this answers is not in the document]
Comment (dhedey) [reply rustfmt.toml:1]: If you want to propose changing the guidelines to add a custom max width we can discuss it as a team

File: scripts/format.sh [hunk 3]
```diff
@@ -1,8 +1,15 @@
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
 
-cd $SCRIPT_DIR/..
-cargo +nightly-2024-02-01 fmt -- --config max_width=120
-cargo +nightly-2024-02-01 fmt -- --config max_width=110
-cargo +nightly-2024-02-01 fmt -- --config max_width=100
-cargo +nightly-2024-02-01 fmt -- --config max_width=90
-cargo +nightly-2024-02-01 fmt -- --config max_width=80
\ No newline at end of file
+# Sadly, rustfmt sometimes can't format lines when they get too long. This could
```
[the remark this answers is not in the document]
Comment (Sajjon) [reply scripts/format.sh:3]: Have you tried https://github.com/dtolnay/prettyplease ?

File: crates/sbor-json/src/common/address.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sajjon) [reply crates/sbor-json/src/common/address.rs:37]: I was suprised to see this too when I fmt some Rust code recently. I think `0. 0` looks super weird :P

File: scripts/format.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhedey) [reply scripts/format.sh:3]: I think prettyplease is mostly intended for formatting generated code rather than formatting user-created code / replacing `rustfmt`

File: Cargo.lock
[the remark this answers is not in the document]
Comment (0xOmarA) [reply Cargo.lock:1]: This is most likely the result of the dependencies being moved into the workspace and me bumping the version on a number of the dependencies.

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/pool_contribution.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            | TakeFromWorktopInstructions::TakeAllFromWorktop(..),
+        ) => true,
+        // Permitted Instructions
+        GroupedInstruction::ReturnToWorktopInstructions(..)
+        | GroupedInstruction::AssertionInstructions(..)
+        | GroupedInstruction::ProofInstructions(..) => true,
+        // Disallowed Instructions
+        GroupedInstruction::TakeFromWorktopInstructions(
+            TakeFromWorktopInstructions::TakeNonFungiblesFromWorktop(..),
+        )
```
Comment (dhedey) [inline crates/radix-engine-toolkit/src/manifest_analysis/classification/pool_contribution.rs:221]: This seems quite arbitrary - to simplify things, can we just allow all `TakeFromWorktopInstructions`; and just restrict the withdraws/deposits to fungibles only?

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+                Self::InitialState,
+                GroupedInstruction::InvocationInstructions(
+                    InvocationInstructions::CallMethod(CallMethod {
+                        address: ManifestGlobalAddress::Static(address),
+                        method_name,
+                        ..
+                    }),
+                ),
+            ) if matches!(
+                address.as_node_id().entity_type().map(Into::into),
```
Comment (dhedey) [inline crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs:101]: MINOR: I'd consider adding an extension method to `pub(crate) trait GlobalAddressExt { grouped_entity_type(&self) -> Option<GroupedEntityType> }`

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                Some(GroupedEntityType::AccountEntities(..))
+            ) && method_name == ACCOUNT_LOCK_FEE_IDENT =>
```
Comment (dhedey) [inline crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs:121]: MINOR: This is repeated a lot, it'd be less noisy to review/understand if this was something like:

```rust
impl GroupedInstruction {
    fn static_only_invocation_matches(
        &self,
        by_entity: impl FnOnce(GroupedEntityType) -> bool,
        by_method: impl FnOnce(&str) -> bool,
    ) -> bool {
        // ...
    }

    fn invocation_matches(
        &self,
        context: AnalysisContext<'_>, // To map dynamic addresses
        by_entity: impl FnOnce(GroupedEntityType) -> bool,
        by_method: impl FnOnce(&str) -> bool,
    ) -> bool {
        // ...
    }
}

impl AnalysisContext {
    fn resolve_entity_type(&self, address: ManifestAddress) -> Option<GroupedEntityType>;
}
```

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/validator_unstake.rs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+            TakeFromWorktopInstructions::TakeFromWorktop(..)
+            | TakeFromWorktopInstructions::TakeAllFromWorktop(..),
+        ) => true,
+        // Permitted Instructions
+        GroupedInstruction::ReturnToWorktopInstructions(..)
+        | GroupedInstruction::AssertionInstructions(..)
+        | GroupedInstruction::ProofInstructions(..) => true,
+        // Disallowed Instructions
+        GroupedInstruction::TakeFromWorktopInstructions(
+            TakeFromWorktopInstructions::TakeNonFungiblesFromWorktop(..),
```
Comment (dhedey) [inline crates/radix-engine-toolkit/src/manifest_analysis/classification/validator_unstake.rs:213]: Ditto here

**Review by dhedey — COMMENTED** [review]
Nice! Didn't notice any logical issues. Lots of tests which are great.

I think most/all of the comments are just a few minor code structure things that I commented on as I passed by, but mostly just personal preference. Although in some cases, we are duplicating quite a lot of code and maybe could make things less boilerplaty with some nice semantic methods.

**Review by dhedey — APPROVED** [review]

File: crates/radix-engine-toolkit/src/manifest_analysis/analysis.rs [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+        // subintents and can't be yielded into.
+        let initial_worktop_state_is_unknown = manifest.is_subintent();
+        let interpreter = StaticManifestInterpreter::new(
+            ValidationRuleset::babylon_equivalent(),
+            manifest,
+        );
+        let mut visitor = StaticResourceMovementsVisitor::new(
+            initial_worktop_state_is_unknown,
+        );
+        interpreter.validate_and_apply_visitor(&mut visitor)?;
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/analysis.rs:53]: 1. All of these functions now return a `Result` and gateway is expected to `try/catch` on them.
2. We're more confident in the underlying static analyzer and typed native invocations. We understand the edge cases and I think that we accept this risk (e.g., if you give me an access rule with a named address I won't be able to convert it into a typed invocation).
3. We have tests now to make sure that typed invocations cover all the cases that we care about.
4. We've pretty much removed the entire old static analyzer backend and replaced it with the new static analyzer. Meaning that the static analyzer will run no matter what. So, exposing a method with and without it is no longer possible.

File: scripts/format.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xOmarA) [reply scripts/format.sh:3]: Looking at `prettyplease` it seems to mention that it's for the output of `syn` so it seems like it's as David says, something more intended for macros.

File: crates/sbor-json/src/common/address.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/sbor-json/src/common/address.rs:37]: Lol agreed, I think that `0. 0` looks very strange. The good news is that it seems like newer versions of Rust moved away from it!

File: crates/radix-engine-toolkit/src/manifest_analysis/types/static_analysis.rs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    /// for. This includes both account and identity entities.
+    pub entities_requiring_auth_summary: EntitiesRequiringAuthOutput,
+
+    /// A summary of the reserved instructions that the toolkit has found in the
+    /// manifest.
+    pub reserved_instructions_summary: ReservedInstructionsOutput,
+
+    /// An ordered set of the classification(s) of the manifest based on the
+    /// static analysis.
+    pub manifest_classification: BTreeSet<ManifestClassification>,
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/types/static_analysis.rs:67]: Resolving this as we discussed it during our call.

File: crates/radix-engine-toolkit/src/manifest_analysis/types/analysis_context.rs [hunk 10]
```diff
@@ -0,0 +1,82 @@
+use crate::internal_prelude::*;
+
+/// Context given by the traverser to the visitors when calling them.
+#[derive(Clone, Copy, Debug)]
+pub enum AnalysisContext<'a> {
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/types/analysis_context.rs:5]: I think that makes sense. I've made this change.

File: crates/radix-engine-toolkit/src/manifest_analysis/traits/requirement_state.rs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    /// manifest traverser were to stop at this instruction then the visitor's
+    /// requirements would be fulfilled and an output would be produced assuming
+    /// that the permission state also resolves to [`true`].
+    Fulfilled,
+    /// In this state the requirements of the visitor are unfulfilled but not
+    /// violated and could certainly be fulfilled if more instructions are seen
+    /// by the visitor. If the manifest were to end with a visitor in this state
+    /// then the visitor would not produce output since it's requirements were't
+    /// fulfilled.
+    CurrentlyUnfulfilled,
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/traits/requirement_state.rs:47]: It not being used was a bit of an oversight. It should be used. The idea behind it is that it's a way for the visitor to signal to the traverser that its requirements can no longer be fulfilled and that there's no point in continuing to traverse the manifest. I've updated the traverser to handle this.

File: crates/radix-engine-toolkit/src/manifest_analysis/requirements/any.rs [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+    fn process_instruction(&mut self, _: AnalysisContext<'_>) {}
+}
+
+macro_rules! define_any_of_requirement_internal {
+    (
+        $($generic: ident),* $(,)?
+    ) => {
+        paste! {
+            impl<$($generic: ManifestAnalyzerRequirementState),*> ManifestAnalyzerRequirementState
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/requirements/any.rs:35]: Thanks! I was super happy with them once I implemented them and saw how they looked in code! I also really like it from a testing and code reuse perspective since it's the same code powering the requirements for all of the classifications.

File: crates/radix-engine-toolkit/src/manifest_analysis/requirements/account_resources_withdrawn_are_not_deposited_back_requirement.rs [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    fn process_instruction(&mut self, context: AnalysisContext<'_>) {
+        if let AnalysisContext::InvocationInstruction {
+            typed_native_invocation:
+                Some(TypedNativeInvocation {
+                    invocation:
+                        TypedManifestNativeInvocation::AccountBlueprintInvocation(
+                            AccountBlueprintInvocation::Method(
+                                AccountBlueprintMethod::Withdraw(
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/requirements/account_resources_withdrawn_are_not_deposited_back_requirement.rs:30]: I feel like this is an area where we could be grouping things a bit too early and we might find ourself ungrouping them later on. As an example, what if we start needing more information than just the resource address in these enum variants? We would now be faced with two options:

1. Add an additional field which is an enum of the amount or resource address.
2. Resort to the typed invocations without the grouping.

It could be that there's a better abstraction/model/definition for this, but I'm kind of waiting for it to emerge :)

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/account_settings_update.rs [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::internal_prelude::*;
+
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+pub struct AccountSettingsUpdateAnalyzer(AccountSettingsUpdateOutput);
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/classification/account_settings_update.rs:21]: Indeed

File: crates/radix-engine-toolkit/src/manifest_analysis/requirements/any.rs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+impl ManifestAnalyzerRequirementState for AnyOfRequirement<()> {
+    fn requirement_state(&self) -> RequirementState {
+        RequirementState::Fulfilled
+    }
+
+    fn process_instruction(&mut self, _: AnalysisContext<'_>) {}
+}
+
+macro_rules! define_any_of_requirement_internal {
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/requirements/any.rs:30]: Sounds good, excited to take a look!

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/pool_contribution.rs (hunk 4 above)
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/classification/pool_contribution.rs:221]: Yeah, this is currently redundant actually since all of the classifications have the requirements needed to ensure that this case of the match statement never get hit. I'm in favor of removing this special casing to just act on groups rather than on individual members.

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs (hunk 5 above)
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs:101]: I think instead of doing that, using typed invocations here would make a lot more sense and we no longer need to do the entity type check here since it's been done for us when converting into the typed invocations.

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs (hunk 6 above)
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/classification/simple_transfer.rs]: Similar to the comment above, I've made a change where we move to use typed invocations which cleans this code up by a lot.

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/validator_claim.rs [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+        else {
+            return;
+        };
+
+        let validator_address = ComponentAddress::try_from(*validator_address)
+            .expect(
+                "Must succeed since the typed invocation conversion succeeded",
+            );
+
+        let input = dynamic_analysis_invocation_io.input.items_iter().next();
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/classification/validator_claim.rs:100]: Yeah I think that it can be improved. I personally like `dynamic_analysis_invocation_io.input.first()` better so that we're removing the iteration and the next call.

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/pool_redemption.rs [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+
+    fn new(
+        _: Self::Initializer,
+    ) -> (Self, Self::PermissionState, Self::RequirementState) {
+        (
+            Default::default(),
+            CallbackPermissionState::new(is_instruction_permitted),
+            Default::default(),
+        )
+    }
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/classification/pool_redemption.rs:45]: > I think if I were tweaking this, I'd get rid of PermissionState as it's not clear to me what it's achieving vs the RequirementState

Permission state is just state that keeps track of what's permitted and if the analyzer should accept more input or not. It's very different in terms of what information captures.

> And internally, have requirements gathering during process_instruction using internal state saved into the analyzer.

I personally want to move away from this and want to have as much structure as possible to the visitors. I think that having this structure allowed for things like the declarative requirements system to emerge. I think that moving to less visitor structure is not the right thing.

File: crates/radix-engine-toolkit/src/manifest_analysis/requirements/any.rs [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+                for AnyOfRequirement<( $($generic,)* )>
+            {
+                fn requirement_state(&self) -> RequirementState {
+                    let (
+                        $(ref [<$generic: snake>],)*
+                    ) = self.0;
+
+                    let requirements = [$([<$generic: snake>].requirement_state()),*];
+
+                    if requirements.iter().any(|r| matches!(r, RequirementState::Fulfilled)) {
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/manifest_analysis/requirements/any.rs:62]: I like this! I've made this change and it feels very nice lol

File: crates/radix-engine-toolkit/src/manifest_analysis/traits/requirement_state.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (dhedey) [reply crates/radix-engine-toolkit/src/manifest_analysis/traits/requirement_state.rs:47]: Are you saying that the traverser is checking this every instruction as a potential performance improvement? I think it's probably of questionable benefit as an optimisation, but I don't mind it.

I imagine it would simplify things to move to a final-outcome based model.

File: crates/radix-engine-toolkit/src/manifest_analysis/classification/pool_redemption.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (dhedey) [reply crates/radix-engine-toolkit/src/manifest_analysis/classification/pool_redemption.rs:45]: But conceptually you can just compose them; rather than having this at the interface layer, and requiring jumping through hoops.

```rust
fn process_instruction(&mut self, context: AnalysisContext<'_>) {
    self.requirements.handle(context);
    // ... other processing
}

fn into_output(self) -> ClassificationResult<Self::Output> {
    self.requirements.result()?;
    // ... actual output
}
```

You could add this boilerplate into a macro or a higher-level type-based abstraction if you wanted; rather than making the `SimpleTransfer` resolver jump through hoops.

And actually I think we'll find that more classifications move the way of state-machines in future anyway, and combine requirements processing with state generation.
